### PR TITLE
Use ESGetTokens in Egamma regression tools

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

After https://github.com/cms-sw/cmssw/pull/31892, this is the next step to migrate the `RecoEgamma/EgammaElectronProducers` package, but it also affects other packages.

#### PR validation:

CMSSW compiles, local matrix tests pass.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport intended.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please test

**Comment by guitargeek** [comment]
Here is what's going on: the object modifier plugins are used both in this module:
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATElectronSlimmer.cc
And in this module:
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/ModifiedObjectProducer.h

The `PATElectronSlimmer` gets the ES products every event, and the ModifiedObjectProducer updates them at every LuminosityBlock transition. The simplest solution here is to also get the ES products at the event transition in the ModifiedObjectProducer. It this a fair solution?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
No reco differences, but there are some DQM differences from memory monitoring. I think it would be good to run a quick profiling for memory usage, just to be sure.
Code review looks fine - I didn't see major items that would make me want to ask for changes. Cases where ConsumesCollector is passed with either & or && were a bit hard to follow, perhaps some comments would be useful.

**Comment by jpata** [comment]
enable profiling

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
@jpata, yeah the thing with the `ConsumesCollector &&` or `&`seems a bit arbitrary. I'll explain how the decisions were made.

Basically it would make sense to use `ConsumesCollector &` by default (you can't use `const &` because the state of the collector is changed). I think the reason why many helpers use the r-value reference `&&` is that this allows to do the plugin-side call to `consumesCollector()` directly in the helper class's constructor. If the constructor would expect a normal reference `&`, you would have to assign the consumes collector first so it has an address, i.e.:
```
auto cc = consumesCollector();
helper_ = Helper{config, cc};
```
However, this is out of the question if `Helper` has do be constructed in the initialization list of the plugin, for example if the Helper has no default constructor `Helper()`.

Now, there are also helper classes that are also used by other helper classes, so the consumes collector gets passed from the top-level helper class to the next one. In this case, I use `&` to avoid usage of `std::move`.

**Comment by jpata** [comment]
Thanks for this explanation, it's very useful!

**Comment by jpata** [comment]
hmm, unless I missed something, the MEM profile was not run by default. Probably it must be turned on somehow, but I'm not sure how now (looking into it).

**Comment by jpata** [comment]
I ran an igprof memory profile myself on 23434.21 with 20 events, I observe no difference.

http://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31999/orig
http://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31999/new

**Comment by jpata** [comment]
+reconstruction

- technical, migrates RecoEgamma/EgammaElectronProducers to ESGetToken
- no changes expected or observed
- no differences observed in CPU or memory profiling, as expected

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1
